### PR TITLE
Implement the JointLimitsTask for the IK

### LINKED ISSUE #530: QPInverseKinematics and joint limit constraints?
Has been using QPInverseKinematics for a while. Usage is similar to IK/tests/QPInverseKinematicsTest.cpp. 

Say I define a CoM task with highpriority and JointTracking task with lowpriority. It happens a lot that CoM is reached but the resulting joint positions violate urdf joint limits.

Is it possible for the solver to take into account joint limit constraint as well? 

Thanks.

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This closes #530 

cc @shikui08

### CODE REVIEW

File: src/IK/src/JointLimitsTask.cpp [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        {
+            if (!joint->getPosLimits(dof,
+                                     m_lowerLimits[joint->getDOFsOffset() + dof],
+                                     m_upperLimits[joint->getDOFsOffset() + dof]))
+            {
+                continue;
+            } else
+            {
+                m_isLimitConsideredForAllJoints = true;
+            }
```
Comment (S-Dafarra) [inline src/IK/src/JointLimitsTask.cpp:57]: It seems that is enough that ``joint->getPosLimits`` returns true for only one joint in order to have ``m_isLimitConsideredForAllJoints = true``. Is this wanted?

File: src/IK/src/JointLimitsTask.cpp [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        return false;
+    }
+
+    // check if the
+    bool useModelLimits = true;
+    if (!ptr->getParameter("use_model_limits", useModelLimits))
+    {
+        log()->error("{} Error while retrieving the 'use_model_limits' parameter", errorPrefix);
+        return false;
+    }
```
Comment (S-Dafarra) [inline src/IK/src/JointLimitsTask.cpp:167]: If there is an error with a specific parameter, I would specify in the message the exact name of the parameter that triggered the error

File: src/IK/src/JointLimitsTask.cpp (hunk 1 above)
Comment (GiulioRomualdi) [reply src/IK/src/JointLimitsTask.cpp:57]: Nope

**Comment by S-Dafarra** [comment]
Thinking about it, does it make sense to have a test case in which we check that the joint limits are actually satisfied over time?

**Comment by shikui08** [comment]
> This closes #530
> 
> cc @shikui08

Wow. Great work. Appreciate your efforts~!

File: src/IK/src/JointLimitsTask.cpp (hunk 2 above)
Comment (GiulioRomualdi) [reply src/IK/src/JointLimitsTask.cpp]: Done 2865e2b6b4fc47582008579d5bee78e4e2877e39

File: src/IK/src/JointLimitsTask.cpp (hunk 1 above)
Comment (GiulioRomualdi) [reply src/IK/src/JointLimitsTask.cpp]: Fixed here 2e8d668ed148b44c4f13dc2a42057d6f81b449fd thank you for noticing

**Comment by GiulioRomualdi** [comment]
> Thinking about it, does it make sense to have a test case in which we check that the joint limits are actually satisfied over time?

This part is currently missing

**Comment by S-Dafarra** [comment]
> > Thinking about it, does it make sense to have a test case in which we check that the joint limits are actually satisfied over time?
> 
> This part is currently missing

Are you planning to add it, or you prefer to avoid it?

**Comment by GiulioRomualdi** [comment]
> Are you planning to add it, or you prefer to avoid it?

I'm trying to do it

**Comment by GiulioRomualdi** [comment]
> > Are you planning to add it, or you prefer to avoid it?
> 
> I'm trying to do it

While writing the test I found a bug that I fixed in https://github.com/ami-iit/bipedal-locomotion-framework/pull/603/commits/4139b0de6403a370b0b871b2a70da3bc2f0f13ef

**Comment by GiulioRomualdi** [comment]
In https://github.com/ami-iit/bipedal-locomotion-framework/pull/603/commits/9fea6654a47d4ca6b9b9ddcc3a9bd87705a63c1d I added a test where I check that the joint limit task is satisfied. I forced the joint to be

$$
q^- \le q \le q^+
$$

where I asked

$$
q^- = q^+ = q(t = 0)
$$

for some joints

**Review by S-Dafarra — APPROVED** [review]

**Pull request merged**